### PR TITLE
Support for Azure OpenAI Realtime

### PR DESCRIPTION
**Pull request by davidzhao**
Ported from Python PR: https://github.com/livekit/agents/pull/848

### CODE REVIEW

File: plugins/openai/src/realtime/api_proto.ts [hunk 1]
```diff
@@ -240,7 +240,7 @@ export interface SessionUpdateEvent extends BaseClientEvent {
     tools: Tool[];
     tool_choice: ToolChoice;
     temperature: number;
-    max_response_output_tokens: number | 'inf';
+    max_response_output_tokens: number | 'inf' | undefined;
```
Comment (nbsp) [inline plugins/openai/src/realtime/api_proto.ts]:
```suggestion
    max_response_output_tokens?: number | 'inf';
```

File: plugins/openai/src/realtime/realtime_model.ts [hunk 2]
```diff
@@ -19,8 +18,11 @@ interface ModelOptions {
   temperature: number;
   maxResponseOutputTokens: number;
   model: api_proto.Model;
-  apiKey: string;
+  apiKey: string | undefined;
```
Comment (nbsp) [inline plugins/openai/src/realtime/realtime_model.ts]:
```suggestion
  apiKey?: string;
```

File: plugins/openai/src/realtime/realtime_model.ts [hunk 3]
```diff
@@ -19,8 +18,11 @@ interface ModelOptions {
...
   baseURL: string;
+  provider: 'openai' | 'microsoft';
```
Comment (nbsp) [inline plugins/openai/src/realtime/realtime_model.ts]: let's make this an enum

File: plugins/openai/src/realtime/realtime_model.ts [hunk 4]
```diff
@@ -446,20 +522,48 @@ export class RealtimeSession extends multimodal.RealtimeSession {
...
-        `${this.#opts.baseURL}?model=${encodeURIComponent(this.#opts.model)}`,
-        {
-          headers: {
-            Authorization: `Bearer ${this.#opts.apiKey}`,
-            'OpenAI-Beta': 'realtime=v1',
-          },
-        },
-      );
+      const headers: Record<string, string> = {
+        'User-Agent': 'LiveKit Agents JS',
```
Comment (nbsp) [inline plugins/openai/src/realtime/realtime_model.ts]: let's follow the [semantics](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/User-Agent) for user-agents, and have something like 

```
Agents-JS/1.0.0 (+https://github.com/livekit/agents-js)
```
doesn't really change anything but if we add one let's at least make it good

**Review by nbsp — CHANGES_REQUESTED** [review]
nitpick pass, will properly test tomorrow

File: plugins/openai/src/realtime/realtime_model.ts (hunk 4 above)
Comment (davidzhao) [reply plugins/openai/src/realtime/realtime_model.ts]: Feel free to add version if you'd like.

File: plugins/openai/src/realtime/realtime_model.ts [hunk 5]
```diff
@@ -183,6 +185,56 @@ export class RealtimeModel extends multimodal.RealtimeModel {
   #defaultOpts: ModelOptions;
   #sessions: RealtimeSession[] = [];
 
+  static withAzure({
```
Comment (bcherry) [inline plugins/openai/src/realtime/realtime_model.ts:188]: added: `static`

File: plugins/openai/src/realtime/realtime_model.ts [hunk 6]
```diff
@@ -446,23 +522,51 @@ export class RealtimeSession extends multimodal.RealtimeSession {
...
+        'User-Agent': 'LiveKit-Agents-JS',
+      };
+      if (this.#opts.provider === 'microsoft') {
+        // Microsoft API has two ways of authentication
+        // 1. Entra token set as `Bearer` token
+        // 2. API key set as `api_key` header (also accepts query string)
+        if (this.#opts.entraToken) {
+          headers.Authorization = `Bearer ${this.#opts.entraToken}`;
+        } else if (this.#opts.apiKey) {
+          headers['api-key'] = this.#opts.apiKey;
```
Comment (bcherry) [inline plugins/openai/src/realtime/realtime_model.ts:539]: fixed `api_key` -> `api-key`

File: examples/src/minimal_assistant.ts [hunk 7]
```diff
@@ -12,9 +12,21 @@ export default defineAgent({
...
+    
+    if (process.env.AZURE_OPENAI_ENDPOINT) {
+      model = openai.realtime.RealtimeModel.withAzure({
+        baseURL: process.env.AZURE_OPENAI_ENDPOINT,
+        azureDeployment: process.env.AZURE_OPENAI_DEPLOYMENT || '',
+        apiKey: process.env.AZURE_OPENAI_API_KEY,
+        entraToken: process.env.AZURE_OPENAI_ENTRA_TOKEN,
+        instructions: 'You are a helpful assistant.',
+      });
+    } else {
```
Comment (bcherry) [inline examples/src/minimal_assistant.ts:27]: added: support in the example

**Review by bcherry — APPROVED** [review]
I tested this and made some updates, detailed below

**Review by nbsp — APPROVED** [review]
lg. this would technically be a minor change, so if you don't mind plugin-openai version being misaligned with agents that's the better way to do it. but also it's a small enough change i don't think it really matters too much

**Pull request merged**